source/agora/tool/main.d

```diff
@@ -0,0 +1,98 @@
+/+
+ dub.sdl:
+ name "sometest"
+ dependency "vibe-d" version="~>0.8"
++/
+
+import agora.common.Types;
+import agora.api.FullNode;
+import agora.common.crypto.Key;
+import agora.common.Serializer;
+import agora.consensus.data.genesis.Test;
+import agora.consensus.data.Transaction;
+import agora.consensus.data.UTXOSetValue;
+import agora.consensus.UTXOSet;
+import agora.utils.Test;
+
+import std.algorithm;
+import std.random;
+import std.range;
+import std.stdio;
+
+import vibe.core.core;
+import vibe.d;
+
+final class MyUTXOSet : UTXOSet
+{
+    public this ()
+    {
+        super(":memory:");
+    }
+
+    public int opApply (scope int delegate (const Hash, const UTXOSetValue) dg)
+    {
+        auto results = this.utxo_db.db.execute("SELECT key, val FROM utxo_map");
+        foreach (row; results)
+        {
+            const key = deserializeFull!Hash(row.peek!(ubyte[])(0));
+            const value = deserializeFull!UTXOSetValue(row.peek!(ubyte[])(1));
+            if (auto ret = dg(key, value))
+                return ret;
+        }
+        return 0;
+    }
```

Comment:
How do we handle this ? We could just move it to the superclass (removing the need to inherit). The current design with the DB being separated doesn't quite fulfill  the open/close principle.